File: ast.py
class Add:
    def __init__(self, left, right):
        self.left = left
        self.right = right
class Mul:
    def __init__(self, left, right):
        self.left = left
        self.right = right
class Num:
    def __init__(self, num):
        self.num = num

def eval(expr):
    if isinstance(expr, Add):
        return eval(expr.left) + eval(expr.right)
    elif isinstance(expr, Mul):
        return eval(expr.left) * eval(expr.right)
    elif isinstance(expr, Num):
        return int(expr.num)

def parse(input):
    pass

expr = Add(
    Num("2"),
    Mul(
	Num("3"),
	Add(
	    Num("4"),
	    Num("5")
	)
    )
)
result = eval(expr) # 29

File: ast.hs
data Ast = Add Ast Ast | Mul Ast Ast | Num String
eval :: Ast -> Int
eval (Add left right) = eval left + eval right
eval (Mul left right) = eval left * eval right
eval (Num num) = read num :: Int
parse :: String -> Ast
parse input = undefined
main :: IO ()
main = do
    let expr = Add (Num "2") (Mul (Num "3") (Add (Num "4") (Num "5")))
    let result = eval expr -- 29
    print result

